flink-core/src/main/java/org/apache/flink/core/plugin/PluginUtils.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.plugin;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * Utility functions for the plugin mechanism.
+ */
+public final class PluginUtils {
```

Comment:
I would merge this and `PluginManagerSingleton` together, or even merge them with `PluginManager`. It's unclear for me what is the benefit of having those two separated and the drawback is that it increases confusion and chances that someone will not find the other when looking for a way how to initialize plugins.

Reply:
The advantage is that some code could be interested in having multiple plugin manager or explicitly creating and passing in plugin managers. The singleton pattern is just a shortcut method that somehow helps avoiding to touch too much code at the cost of having kind of a global variable. Therefore I kept a separation. If you still feel like the merge is better, I would be ok with that as well.

Reply:
And for this particular utils class, if we want to keep plugin management decoupled from the discovery mechanism (directory based, config based,...) then it feels wrong to have this code in the singleton class. The current init in the singleton is independent from the discovery mechanics.

Reply:
I also agree with @pnowojski to merge it into `PluginManager`. From a point viewer of plugin user, I would expect there's one only entry point to use plugin. So basically, I suppose it involves following 2 steps:
1. init PluginManager
2. create Plugin via PluginManager

> some code could be interested in having multiple plugin manager or explicitly creating and passing in plugin managers.

I am sure why user would do that. Only one scenario I can imagine is that user want to create a customized PluginManager explicitly for unit test. And I think the only thing to customize PluginManager is its plugin root folder, this could be set by a lot approaches like java property or system environment, so I think PluginManger itself could do that, we don't need PluginUtil for this. Maybe my understanding is wrong, just my 2 cents. 😄